anvill/python/anvill/binja.py

```diff
@@ -173,202 +173,253 @@ def _convert_bn_llil_type(
         ret = IntegerType(reg_size_bytes, True)
         return ret
 
-def _cache_key(tinfo: bn.types.Type) -> str:
-    """ Convert bn Type instance to cache key"""
-    return str(tinfo)
-
-def _convert_named_type_reference(
-    bv, tinfo: bn.types.Type, cache) -> Type:
-    """ Convert named type references into a `Type` instance"""
-    if tinfo.type_class != bn.TypeClass.NamedTypeReferenceClass:
-        return
 
-    named_tinfo = tinfo.named_type_reference
-    if (named_tinfo.type_class
-        == bn.NamedTypeReferenceClass.StructNamedTypeClass):
-        # Get the bn struct type and recursively recover the elements
-        ref_type = bv.get_type_by_name(named_tinfo.name);
-        struct_type =  ref_type.structure
+def get_arch(bv):
+    """Arch class that gives access to architecture-specific functionality."""
+    name = bv.arch.name
+    if name == "x86_64":
+        return AMD64Arch()
+    elif name == "x86":
+        return X86Arch()
+    elif name == "aarch64":
+        return AArch64Arch()
+    else:
+        raise UnhandledArchitectureType(
+            "Missing architecture object type for architecture '{}'".format(name)
+        )
+
+
+def get_os(bv):
+    """OS class that gives access to OS-specific functionality."""
+    platform = str(bv.platform)
+    if "linux" in platform:
+        return LinuxOS()
+    elif "mac" in platform:
+        return MacOS()
+    elif "windows" in platform:
+        return WindowsOS()
+    else:
+        raise UnhandledOSException(
+            "Missing operating system object type for OS '{}'".format(platform)
+        )
+
+
+class TypeCache:
+    """The class provides API to recursively visit the binja types and convert
+    them to the anvill `Type` instance. It maintains a cache of visited binja
+    types to reduce lookup time.
+    """
+
+    __slots__ = ("_bv", "_cache")
+
+    # list of unhandled type classes which should log error
+    _err_type_class = {
+        bn.TypeClass.VarArgsTypeClass: "VarArgsTypeClass",
+        bn.TypeClass.ValueTypeClass: "ValueTypeClass",
+        bn.TypeClass.WideCharTypeClass: "WideCharTypeClass",
+    }
+
+    def __init__(self, bv):
+        self._bv = bv
+        self._cache = dict()
+
+    def _cache_key(self, tinfo: bn.types.Type):
+        """ Convert bn Type instance to cache key"""
+        return str(tinfo)
+
+    def _convert_struct(self, tinfo: bn.types.Type):
+        """Convert bn struct type into a `Type` instance"""
+
+        assert tinfo.type_class == bn.TypeClass.StructureTypeClass
+
+        if tinfo.structure.type == bn.StructureType.UnionStructureType:
+            return self._convert_union(tinfo)
+
+        assert tinfo.structure.type == bn.StructureType.StructStructureType
```

Comment:
update assert to check for `ClassStructureType`

Reply:
Thanks, for pointing. `ClassStructureType` was not getting handled. It will be the same as `StructType()`